### ISSUE TITLE
[jvm-packages] Can't actually build the jvm-package without "-DskipTests" option

### ISSUE THREAD

**Issue by avloss** [dmlc/xgboost#1105]
Actions (using Maven 3.3.9):
```
git clone --recursive https://github.com/dmlc/xgboost
cd xgboost/jvm-packages
mvn packeage
```

expected:
package builds

actual:
build gets stuck with the following lines 
```
16/04/12 03:27:19 INFO TaskSchedulerImpl: Removed TaskSet 0.0, whose tasks have all completed, from pool
16/04/12 03:27:19 INFO DAGScheduler: Job 0 finished: foreachPartition at XGBoost.scala:104, took 0.090704 s
```

when checking just stderr, following messages can be seen
```
Apr 12, 2016 3:28:08 AM ml.dmlc.xgboost4j.java.XGBoost crossValidation
INFO: [0]	cv-train-error:0.014431	cv-test-error:0.014439
Apr 12, 2016 3:28:08 AM ml.dmlc.xgboost4j.java.XGBoost crossValidation
INFO: [1]	cv-train-error:0.001612	cv-test-error:0.001843
```
I assumed that this could be because some special set-up on the developer machine is needed to pass all the unit tests, but as advised by @CodingCat - tests should pass on every machine. 
I'm using: 
instace: c4.4xlarge AWS
kernel: 3.13.0-74-generic x86_64
os: Ubuntu 14.04.3 LTS

**Comment by CodingCat** [comment]
@avloss thanks for creating the ticket!

Would you attach the complete log trace here, I would like to know which test case is the last successfully finished one.....I think the potential issue would be that the tracker is now started properly

**Comment by avloss** [comment]
03:46:21 is the time when (after waiting for 5 minutes and seeing nothing happening, CPUs seemed idle) I've aborted the build

[mvn_package.txt](https://github.com/dmlc/xgboost/files/214289/mvn_package.txt)

**Comment by CodingCat** [comment]
Hey, @avloss I went through the log file, it seems that the workers (distributed process training the model) does not connect back to the tracker (controller of the program) properly 

in the last part of the log, I haven't seen the lines like 

```
16/04/12 20:19:37 INFO RabitTracker$TrackerProcessLogger: 2016-04-12 20:19:37,427 DEBUG Recieve start signal from 192.168.0.196; assign rank 0
16/04/12 20:19:37 INFO RabitTracker$TrackerProcessLogger: 2016-04-12 20:19:37,427 DEBUG Recieve start signal from 192.168.0.196; assign rank 1
16/04/12 20:19:37 INFO RabitTracker$TrackerProcessLogger: 2016-04-12 20:19:37,428 DEBUG Recieve start signal from 192.168.0.196; assign rank 2
16/04/12 20:19:37 INFO RabitTracker$TrackerProcessLogger: 2016-04-12 20:19:37,431 DEBUG Recieve start signal from 192.168.0.196; assign rank 3

```

(The above lines are copied from my own laptop when running the tests) 

There might be some network configuration issue in AWS instance, causing that the workers cannot connect your local IP address, 

@tqchen when you ran the benchmark in AWS, did you meet some similar issues on networking?

**Comment by avloss** [comment]
to me this sounds like certain network profile is required for this build to run. Do you expect ports on "192.168.0.196" to be open? Even though this is a local interface, I might chose to have certain FW rules on it. Your build should either create some kind of container, over which it'll be able to control firewall settings, or alternatively, add that "skipTests" flag to the builds for "non-develpers".

TL;DR
- Package DOES work after building with `skipTests`
- There must be some FW settings your functional tests do not like
- Having ports open on the local interface for the build to complete successfully is too much of a requirement. Do you expect people to update their security settings just so build can run it's functional tests?

**Comment by CodingCat** [comment]
I cannot understand how you run the unit tests on machine A and start tracker on another machine

**Comment by CodingCat** [comment]
and also, if you cannot bind to 192.168.0.196, that means once you run XGBoost.train, the program will stuck...I do not have the idea  about how the package would work...

**Comment by CodingCat** [comment]
in summary, 

* since the building process is done with maven, feel free to add any options which you prefer when doing your own work, 

* I will keep `-DskipTests` in the document to prevent any incorrect configuration in their local environment

**Comment by avloss** [comment]
I've just re-run the build with tests - and it worked. So I can't reproduct the issue anymore. Closing.

**Comment by sloth2012** [comment]
>Actually, I met the same question. Would you tell me how do you fix it?

**Comment by avloss** [comment]
try doing the build in a different folder - as I recall - if you do build incorrectly the first time, it leaven some issues behind. So just start in a new folder (do `git clone` again)!

**Comment by sloth2012** [comment]
I tried, but failed every time. I can't understand how it happened like this.

**Comment by sloth2012** [comment]
Finally, I fix it. I resolve it by reinstall gcc via brew, as my env is mac osx LLVM gcc.

**Comment by Sinan81** [comment]
What version of gcc did you use?
